### PR TITLE
Migrate Continuous Aggregates to the new format

### LINKED ISSUE #4424: Migrate continuous aggregates from partial form to final form
# Goals

Provide a method to upgrade existing continuous aggregates to from the old format to the new format. The method should be usable also for very large amounts of data and support methods to gradually upgrade to the new format.

# Background

Continuous aggregates store the intermediate result in a *materialization table* and in that sense behave similar to materialized views with an automated update. The aggregation state was originally stored as a byte array of the [partial aggregate](https://www.postgresql.org/docs/current/xaggr.html#XAGGR-PARTIAL-AGGREGATES) form, which is an internal form and as such did not have a stable and exportable format.

With #4269 this was changed so that aggregates are stored in final form instead, offering multiple opportunities for optimizations as well as a more compact format. When upgrading to Timescale 2.7, new continuous aggregates are using the new format, but existing continuous aggregates keep using the format they were defined with.

# Problem

Users that already have continuous aggregate and wish to use the new format have no means to upgrade existing continuous aggregates to the new format.

### PR DESCRIPTION
**Pull request by fabriziomello**
Timescale 2.7 released a new version of Continuous Aggregate (#4269)
that store the final aggregation state instead of the byte array of
the partial aggregate state, offering multiple opportunities of
optimizations as well a more compact form.

When upgrading to Timescale 2.7, new created Continuous Aggregates
are using the new format, but existing Continuous Aggregates keep
using the format they were defined with.

Created a procedure to upgrade existing Continuous Aggregates from
the old format to the new format, by calling a simple procedure:

```sql
test=# CALL cagg_migrate('conditions_summary_daily');
```

Closes #4424

### CODE REVIEW

File: sql/updates/latest-dev.sql [hunk 1]
```diff
@@ -21,3 +21,39 @@ ALTER TABLE _timescaledb_catalog.chunk
...
+);
+
+SELECT pg_catalog.pg_extension_config_dump('_timescaledb_catalog.continuous_agg_migrate_plan_step', '');
+
+SELECT pg_catalog.pg_extension_config_dump(pg_get_serial_sequence('_timescaledb_catalog.continuous_agg_migrate_plan_step', 'step_id'), '');
+
+-- in tables.sql the same is done with GRANT SELECT ON ALL TABLES IN SCHEMA
+GRANT SELECT ON _timescaledb_catalog.continuous_agg_migrate_plan TO PUBLIC;
+GRANT SELECT ON _timescaledb_catalog.continuous_agg_migrate_plan_step TO PUBLIC;
+GRANT SELECT ON _timescaledb_catalog.continuous_agg_migrate_plan_step_step_id_seq TO PUBLIC;
```
Comment (mfundul) [inline sql/updates/latest-dev.sql:59]:
```suggestion
GRANT SELECT ON pg_get_serial_sequence('_timescaledb_catalog.continuous_agg_migrate_plan_step', 'step_id') TO PUBLIC;
```
I don't know if we can do this, or if it is even necessary to do the GRANT on the serial. In any case, if we have to use the name `continuous_agg_migrate_plan_step_step_id_seq` anyway, why do we use `pg_get_serial_sequence('_timescaledb_catalog.continuous_agg_migrate_plan_step', 'step_id')` in the first place?

File: sql/updates/reverse-dev.sql [hunk 2]
```diff
@@ -152,3 +152,23 @@ GRANT SELECT ON _timescaledb_catalog.chunk TO PUBLIC;
...
+DROP PROCEDURE _timescaledb_internal.cagg_migrate_execute_enable_policies (_timescaledb_catalog.continuous_agg, _timescaledb_catalog.continuous_agg_migrate_plan_step);
+DROP PROCEDURE _timescaledb_internal.cagg_migrate_execute_copy_policies (_timescaledb_catalog.continuous_agg, _timescaledb_catalog.continuous_agg_migrate_plan_step);
+DROP PROCEDURE _timescaledb_internal.cagg_migrate_execute_refresh_new_cagg (_timescaledb_catalog.continuous_agg, _timescaledb_catalog.continuous_agg_migrate_plan_step);
+DROP PROCEDURE _timescaledb_internal.cagg_migrate_execute_copy_data (_timescaledb_catalog.continuous_agg, _timescaledb_catalog.continuous_agg_migrate_plan_step);
+DROP PROCEDURE _timescaledb_internal.cagg_migrate_execute_override_cagg (_timescaledb_catalog.continuous_agg, _timescaledb_catalog.continuous_agg_migrate_plan_step);
+DROP PROCEDURE _timescaledb_internal.cagg_migrate_execute_drop_old_cagg (_timescaledb_catalog.continuous_agg, _timescaledb_catalog.continuous_agg_migrate_plan_step);
+
+ALTER EXTENSION timescaledb DROP SEQUENCE _timescaledb_catalog.continuous_agg_migrate_plan_step_step_id_seq;
```
Comment (mfundul) [inline sql/updates/reverse-dev.sql:170]:
```suggestion
ALTER EXTENSION timescaledb DROP SEQUENCE pg_get_serial_sequence('_timescaledb_catalog.continuous_agg_migrate_plan_step', 'step_id');
```
Do we need to drop the serial if we drop the table? We probably don't need to use `pg_get_serial_sequence()` at all if we have to use `_timescaledb_catalog.continuous_agg_migrate_plan_step_step_id_seq` anyway.

File: sql/updates/latest-dev.sql (hunk 1 above)
Comment (fabriziomello) [reply sql/updates/latest-dev.sql:59]: The syntax you're proposing is not allowed in Postgres, you should specify the object name in GRANT statement: https://www.postgresql.org/docs/current/sql-grant.html

Have a look if we try to run what you're proposing:

```sql
fabrizio=# GRANT SELECT ON pg_get_serial_sequence('_timescaledb_catalog.continuous_agg_migrate_plan_step', 'step_id') TO PUBLIC;
ERROR:  syntax error at or near "("
LINE 1: GRANT SELECT ON pg_get_serial_sequence('_timescaledb_catalog...
```

File: sql/updates/reverse-dev.sql (hunk 2 above)
Comment (fabriziomello) [reply sql/updates/reverse-dev.sql:170]: Same here:

```sql
fabrizio=# ALTER EXTENSION timescaledb DROP SEQUENCE pg_get_serial_sequence('_timescaledb_catalog.continuous_agg_migrate_plan_step', 'step_id');
ERROR:  syntax error at or near "("
LINE 1: ...N timescaledb DROP SEQUENCE pg_get_serial_sequence('_timesca...
                                                             ^
```

**Review by RafiaSabih — APPROVED** [review]

**Review by svenklemm — APPROVED** [review]

File: sql/updates/latest-dev.sql (hunk 1 above)
Comment (mfundul) [reply sql/updates/latest-dev.sql:59]: > In any case, if we have to use the name continuous_agg_migrate_plan_step_step_id_seq anyway

That's why then we should be consistent, if we use `continuous_agg_migrate_plan_step_step_id_seq` we have to remove `pg_get_serial_sequence('_timescaledb_catalog.continuous_agg_migrate_plan_step', 'step_id')`

File: sql/updates/reverse-dev.sql (hunk 2 above)
Comment (mfundul) [reply sql/updates/reverse-dev.sql:170]: > We probably don't need to use pg_get_serial_sequence() at all 

That's why then we should be consistent, if we use `continuous_agg_migrate_plan_step_step_id_seq` we have to remove `pg_get_serial_sequence('_timescaledb_catalog.continuous_agg_migrate_plan_step', 'step_id')` from the code.

File: sql/updates/latest-dev.sql (hunk 1 above)
Comment (fabriziomello) [reply sql/updates/latest-dev.sql:59]: Make sense... fixed!

File: sql/updates/reverse-dev.sql (hunk 2 above)
Comment (fabriziomello) [reply sql/updates/reverse-dev.sql:170]: Fixed

**Pull request merged**